Add support for threading on MacOS

aykevl opened the pull request:
This is based on #4559. This PR itself is very small, most of the work was in making all headers available in lib/macos-minimal-sdk.

aykevl committed:
darwin: add threading support and use it by default

aykevl commented:
This PR is now ready!

eliasnaur commented on src/internal/task/task_threads.c:
```diff
@@ -22,7 +30,11 @@ struct state_pass {
     void      *args;
     void      *task;
     uintptr_t *stackTop;
+    #if __APPLE__
```
Preprocessor directives are somewhat ugly, and get uglier with the addition of platforms (e.g. Windows). I suggest adding a GOOS-specific typedef for the semaphore type along with a handful of small inline functions that abstracts the semaphore API.

aykevl replied:
That is what I originally tried, but in my opinion didn't result in code that was easier to read. Note that the way these functions are called is very different between standard POSIX (`sem_*` API) and MacOS.

I will likely make a new file for Windows. Maybe I can use the pthread library of MinGW and reuse this code, but I expect Windows will be so different a separate file makes things easier. We'll see.

eliasnaur reviewed (approved):
Change looks good. Small suggestion below.

deadprogram commented:
@aykevl something not quite right with this PR, I think. The CI tests on macOS run "forever" until they timeout.

aykevl commented:
Yeah that doesn't look quite right, I need to debug this on MacOS.

I've cancelled the CI for now, it's not going to get any further this way.

aykevl committed:
internal/task: a few small correctness fixes

This shouldn't affect Linux or MacOS, but it's good to have them fixed.

aykevl commented:
Fixed the CI hang: there was a problem with the stop-the-world phase (other threads never received the signal) that is now fixed.

debug/dwarf still fails when I run it locally, but it also fails on the dev branch so that seems to be an unrelated problem. Other than that, stdlib packages now pass the tests.

eliasnaur commented:
Still LGTM.

deadprogram commented:
Thanks for working on this @aykevl and to @eliasnaur for reviews. Now merging.
